server/lib/chat/chat-service.ts

```diff
@@ -118,85 +156,128 @@ export default class ChatService {
+      this.updateUserAfterJoining(userInfo, channelInfo, userChannelEntry, message),
+    )
+  }
+
+  /**
+   * Joins `channelName` with account `userId`, allowing them to receive and send messages in it.
+   * Handles the use case of two users attempting to join a channel at the same time.
+   */
+  async joinChannel(channelName: string, userId: SbUserId): Promise<ChannelInfo> {
+    const userInfo = await findUserById(userId)
+    if (!userInfo) {
+      throw new ChatServiceError(ChatServiceErrorCode.UserNotFound, "User doesn't exist")
+    }
 
-      // NOTE(tec27): We don't use the helper method here because joining channels while offline
-      // is allowed in some cases (e.g. during account creation)
-      const userSockets = this.userSocketsManager.getById(userId)
-      if (userSockets) {
-        this.subscribeUserToChannel(
-          userSockets,
-          userChannelEntry.channelName,
-          channelInfo,
-          userChannelEntry.channelPermissions,
-        )
+    let succeeded = false
+    let attempts = 0
+    let channel: ChannelInfo | undefined
+    let userChannelEntry: UserChannelEntry | undefined
+    do {
+      attempts += 1
+      try {
+        await transact(async client => {
+          channel = await findChannelByName(channelName, client)
+          if (channel) {
+            if (this.state.users.has(userId) && this.state.users.get(userId)!.has(channel.id)) {
+              throw new ChatServiceError(
+                ChatServiceErrorCode.AlreadyJoined,
+                'Already in this channel',
+              )
+            }
+
+            const isBanned = await isUserBannedFromChannel(channel.id, userId, client)
+            if (isBanned) {
+              throw new ChatServiceError(
+                ChatServiceErrorCode.UserBanned,
+                'This user has been banned from this chat channel',
+              )
+            }
+
+            try {
+              userChannelEntry = await addUserToChannel(userId, channel.id, client)
+              succeeded = true
+            } catch (err: any) {
+              if (err.code === FOREIGN_KEY_VIOLATION) {
+                throw new RetryableError()
+              } else {
+                throw err
+              }
+            }
+          } else {
+            try {
+              channel = await createChannel(userId, channelName, client)
+              userChannelEntry = await addUserToChannel(userId, channel.id, client)
+              succeeded = true
+            } catch (err: any) {
+              if (err.code === UNIQUE_VIOLATION) {
+                throw new RetryableError()
+              } else {
+                throw err
+              }
+            }
+          }
+        })
+      } catch (err) {
+        if (!(err instanceof RetryableError)) {
+          throw err
+        }
       }
+    } while (!succeeded && attempts < MAX_JOIN_ATTEMPTS)
+
+    const message = await addMessageToChannel(userId, channel!.id, {
```

Comment:
This isn't safe, you never checked that the preceding operation actually succeeded? If it did not, you need to throw some other error before trying to do all this other stuff

Reply:
Honestly this should almost certainly be done in the transaction (try/catches) as well. I guess you were trying to pull it out to avoid duplicating code but like, duplicated but correct code is a lot better than de-duplicated but incorrect code.